### PR TITLE
WIP-better unload records fixes #3296

### LINKED ISSUE #3296: Unload issue with relationships
It is not clear what unloading should do with regards to relationships. It seems to me that it should dematerialize a record, but keep the internalModel intact if it has SSOT foreign key references(removing foreign keys seems obviously incorrect).
This is problematic though, because internalModels will still leak memory. How do we let users unload a bunch of related records? Does this mean we have to write a mini Garbage Collector, that will anytime you unload a record, check the foreign key dependencies and unload things it can. This doesn't sound too bad, can you confirm it is the right thing to do?

### PR DESCRIPTION
**Pull request by igorT**
@sly7-7

### CODE REVIEW

File: packages/ember-data/lib/system/model/internal-model.js [hunk 1]
```diff
@@ -173,8 +181,63 @@ InternalModel.prototype = {
     return this.record;
   },
 
+  directlyRelatedRecords: function() {
```
Comment (sly7-7) [inline packages/ember-data/lib/system/model/internal-model.js]: maybe s/record/internalModel ?

Comment (igorT) [inline packages/ember-data/lib/system/model/internal-model.js]: please PR to this branch

**Comment by igorT** [comment]
Fixes https://github.com/emberjs/data/issues/3296

**Comment by raytiley** [comment]
@igorT is this still the approach you had in mind?

**Comment by bmac** [comment]
@igorT do you have some time to write up what needs to happen before this pr can be merged?

**Comment by bmac** [comment]
@igorT do you have some time to write up what needs to happen before this pr can be merged?

**Comment by bmac** [comment]
I'm going to close this issue since there hasn't been any response in a while. Feel free to reopen if there is an update.

**Comment by hjdivad** [comment]
@igorT rebased at https://github.com/hjdivad/data/tree/better-unload

I'll try to find some time to finish this

**Pull request closed without merging**